### PR TITLE
Added Countries Translations (Installation & LocalizationPack)

### LINKED ISSUE #12168: Translation of Countries in PS1.7 - installation step
**Describe the bug**
During the installation of PrestaShop, the `日本語` is selected; in Step 4: "Store information" => all Countries are displayed in English.
In fact, they are always displayed in English in any language selected.

**To Reproduce**
1. Install PrestaShop with any language different from English
2. In the in Step 4: "Store information" => all Countries are displayed in English

**Screenshots**
image

**Additionnal information**
PrestaShop version: 1.7.5.0 / 1.7.4.4
PHP version: 7.1

### LINKED ISSUE #18929: Wrong country names ps_country_lang
After adding diffirent languages all Country names are in English.

In DB table ps_country_lang I see this
image

- Installed PS 1.7.6.4 English version with Russian localization
- later languages were added in this sequance as in picture.  
image

#### Additional information

* PrestaShop version: 1.7.6.4 English version with Russian localization
* PHP version: Apache/2.4.33 (Win32) OpenSSL/1.1.0h PHP/7.2.6

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Added Countries Translations in Installation Process
| Type?         | improvement
| Category?     | IN
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #12168 & Fixes #18929
| How to test? | * Start the installation process<br>* Choose a language (Arabic Language)<br>* **BEFORE:** The countries list must be in English<br>* **AFTER:** The countries list must be translated (Fixes #12168)<br>* Choose the United States country for example<br>* Go to BO > Catalog > Brands page > Addresses section<br>* The country is displayed in Arabic<br>* Change the employee language to "English"<br>* Go to BO > Catalog > Brands page > Addresses section<br>* **The country is displayed in English** (Fixes https://github.com/PrestaShop/PrestaShop/pull/19818#issuecomment-704930659)<br>* Check for another pages : Orders page, Addresses page, Countries page<br>* In DB, the table ps_country_lang has countries in English and Arabic<br>* In BO, add the localization pack for Spain<br>* **In DB, the table ps_country_lang has countries in Spanish, English and Arabic** (Fixes #18929)<br>

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/19818)

### CODE REVIEW

**Comment by LouiseBonnard** [comment]
Great, thanks @Progi1984! In what language will the country name be translated with this PR: the language of installation?

**Comment by Progi1984** [comment]
@LouiseBonnard Affirmative.

File: src/Core/Localization/CLDR/LocaleData.php [hunk 1]
```diff
@@ -368,4 +383,24 @@ public function setCurrencies($currencies)
...
+    {
+        return $this->territories;
+    }
+
+    /**
+     * @param string[] $territories
+     *
+     * @return LocaleData
+     */
+    public function setTerritories($territories)
```
Comment (jolelievre) [inline src/Core/Localization/CLDR/LocaleData.php]:
```suggestion
    public function setTerritories(array $territories): LocaleData
```

File: src/Core/Localization/CLDR/LocaleData.php [hunk 2]
```diff
@@ -368,4 +383,24 @@ public function setCurrencies($currencies)
 
         return $this;
     }
+
+    /**
+     * @return string[]
```
Comment (jolelievre) [inline src/Core/Localization/CLDR/LocaleData.php]: Maybe add a comment to define the format, from the Reader code it seems to be an array indexed by `type` ? What does the type attribute refer to?

File: src/PrestaShopBundle/Command/UpdateCountriesCommand.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $container = $this->getContainer();
+        $reader = $container->get('prestashop.core.localization.cldr.reader');
+
+        $template = $input->getArgument('installFolder') . DIRECTORY_SEPARATOR . 'langs' . DIRECTORY_SEPARATOR . 'en'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Command/UpdateCountriesCommand.php]: Only in english?

File: src/PrestaShopBundle/Command/UpdateCountriesCommand.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            throw new Exception(sprintf('File "%s" not found', $template));
+        }
+        $xmlTemplate = new SimpleXMLElement(file_get_contents($template));
+
+        $finder = new Finder();
+        $finder
+            ->directories()
+            ->depth(0)
+            ->in($input->getArgument('installFolder') . DIRECTORY_SEPARATOR . 'langs')
+            ->exclude('en');
```
Comment (jolelievre) [inline src/PrestaShopBundle/Command/UpdateCountriesCommand.php]: If I get it right, the `en` file is used as a base to loop though the countries and update other languages
But then how is the en language itself updated?

File: src/PrestaShopBundle/Command/UpdateCountriesCommand.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $container = $this->getContainer();
+        $reader = $container->get('prestashop.core.localization.cldr.reader');
+
+        $template = $input->getArgument('installFolder') . DIRECTORY_SEPARATOR . 'langs' . DIRECTORY_SEPARATOR . 'en'
+            . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR . 'country.xml';
+        if (!is_file($template)) {
+            throw new Exception(sprintf('File "%s" not found', $template));
```
Comment (PululuK) [inline src/PrestaShopBundle/Command/UpdateCountriesCommand.php]: Use appropriate exception
See https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Exception/FileNotFoundException.php#L29

Comment (Progi1984) [reply src/PrestaShopBundle/Command/UpdateCountriesCommand.php]: @PululuK Finally, I changed the behavior as discussed with my colleagues and use `symfony/intl`

**Review by PierreRambaud — DISMISSED** [review]

**Comment by Progi1984** [comment]
Remove 3 countries because unhabited : 
* AN
* BV
* HM

Reference : https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Intl/Data/Generator/RegionDataGenerator.php#L36-L39

File: src/PrestaShopBundle/Install/XmlLoader.php [hunk 6]
```diff
@@ -367,6 +367,56 @@ protected function getFallBackToDefaultEntityLanguage($iso, $entity)
...
 
+    /**
+     * Special case for "country" entity.
+     */
+    public function populateEntityCountry()
+    {
+        $xml = $this->loadEntity('country');
+
+        // Read list of fields
+        if (!is_object($xml) || !$xml->fields) {
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Install/XmlLoader.php]:
```suggestion
        if (empty($xml->fields)) {
```
Should be enough?

Comment (Progi1984) [reply src/PrestaShopBundle/Install/XmlLoader.php]: I say yes !

File: src/PrestaShopBundle/Install/XmlLoader.php [hunk 7]
```diff
@@ -367,6 +367,56 @@ protected function getFallBackToDefaultEntityLanguage($iso, $entity)
...
+        $langs = [];
+        $languageList = LanguageList::getInstance();
+        foreach ($this->languages as $id_lang => $iso) {
+            $languageList->setLanguage($iso);
+            $langs[$id_lang] = $languageList->getCountries();
+        }
+
+        // Load all row for current entity and prepare data to be populated
+        $i = 0;
+        foreach ($xml->entities->country as $node) {
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Install/XmlLoader.php]: If you test `$xml->fields` should you test `$xml->entities->country` and be sure it's an array, no?

Comment (Progi1984) [reply src/PrestaShopBundle/Install/XmlLoader.php]: Done ;)

**Comment by Progi1984** [comment]
> Remove 3 countries because unhabited :
> 
>     * AN
> 
>     * BV
> 
>     * HM
> 
> 
> Reference : https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Intl/Data/Generator/RegionDataGenerator.php#L36-L39

Ping @LouiseBonnard

**Comment by LouiseBonnard** [comment]
Alright, thanks @Progi1984! Any clue why those 3 inhabited countries were mentioned in the code?

**Comment by Progi1984** [comment]
@LouiseBonnard There are here since 9 years..... It's historical.

File: src/PrestaShopBundle/Install/XmlLoader.php [hunk 8]
```diff
@@ -367,6 +368,58 @@ protected function getFallBackToDefaultEntityLanguage($iso, $entity)
...
+    /**
+     * Special case for "country" entity.
+     */
+    public function populateEntityCountry()
+    {
+        $xml = $this->loadEntity('country');
+
+        // Read list of fields
+        if (empty($xml->fields)) {
+            throw new PrestashopInstallerException('List of fields not found for entity coutrny');
```
Comment (atomiix) [inline src/PrestaShopBundle/Install/XmlLoader.php]:
```suggestion
            throw new PrestashopInstallerException('List of fields not found for entity country');
```

**Review by atomiix — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,

The countries are well translated, but the states are still in English.
For example:
image

Also, after fresh install "Arabic language" and the United States country
Go to the FO => Add new Address, select the English language
All countries are displayed in Arabic, they should be translated, is it?
image

Thanks!

**Comment by Progi1984** [comment]
> Hi @Progi1984,
> 
> The countries are well translated, but the states are still in English.
> For example:
> image

@khouloudbelguith It's not states. It is the timezone. I don't think it's in the scope of this PR.

> Also, after fresh install "Arabic language" and the United States country
> Go to the FO => Add new Address, select the English language
> All countries are displayed in Arabic, they should be translated, is it?
> image
> 
> Thanks!

I didn't reproduce in local :(

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,

Thanks for your feedback.
Waiting for Developper's review to do a new test.

Thanks!

**Comment by Progi1984** [comment]
@khouloudbelguith I reproduce finally the second bug. I'm on it !

**Review by PierreRambaud — DISMISSED** [review]

File: src/PrestaShopBundle/Install/XmlLoader.php (hunk 8 above)
Comment (atomiix) [reply src/PrestaShopBundle/Install/XmlLoader.php]: Still there 😅

Comment (Progi1984) [reply src/PrestaShopBundle/Install/XmlLoader.php]: Fixed

File: install-dev/data/xml/address_format.xml [hunk 9]
```diff
@@ -1054,16 +1054,6 @@ address1
 address2
 postcode city
 Country:name
-phone</format>
```
Comment (jolelievre) [inline install-dev/data/xml/address_format.xml:1057]: The f*** is this?!!

Comment (jolelievre) [reply install-dev/data/xml/address_format.xml:1057]: Some appended format for all countries that need to be split every 9 lines?! Or every `Country:name` instance?

File: src/PrestaShopBundle/Install/LanguageList.php [hunk 10]
```diff
@@ -149,8 +149,13 @@ public function getLanguages()
     public function getCountries()
     {
         static $countries = null;
+        static $language = null;
```
Comment (jolelievre) [inline src/PrestaShopBundle/Install/LanguageList.php]: Well that's not scary at all! 😅

File: src/PrestaShopBundle/Install/XmlLoader.php [hunk 11]
```diff
@@ -367,6 +368,58 @@ protected function getFallBackToDefaultEntityLanguage($iso, $entity)
...
+        $xml = $this->loadEntity('country');
+
+        // Read list of fields
+        if (empty($xml->fields)) {
+            throw new PrestashopInstallerException('List of fields not found for entity country');
+        }
+        $langs = [];
+        $languageList = LanguageList::getInstance();
+        foreach ($this->languages as $id_lang => $iso) {
+            $languageList->setLanguage($iso);
```
Comment (jolelievre) [inline src/PrestaShopBundle/Install/XmlLoader.php]: I see what you're doing with the static variables reset and all But wouldn't it be cleaner and more stable to add a `getCountriesByLanguage` method instead in `LanguageList` ?

File: install-dev/data/xml/address_format.xml (hunk 9 above)
Comment (Progi1984) [reply install-dev/data/xml/address_format.xml:1057]: I remove an address format. Normal ;)

File: src/PrestaShopBundle/Install/XmlLoader.php [hunk 12]
```diff
@@ -367,6 +368,57 @@ protected function getFallBackToDefaultEntityLanguage($iso, $entity)
...
+
+        // Load all row for current entity and prepare data to be populated
+        $i = 0;
+        if ($xml->entities->country instanceof SimpleXMLElement) {
+            foreach ($xml->entities->country as $node) {
+                $data = [];
+                $identifier = (string) $node['id'];
+
+                // Read attributes
+                foreach ($node->attributes() as $k => $v) {
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Install/XmlLoader.php:396]: use xpath maybe?
https://www.php.net/manual/en/domxpath.query.php
```
$xpath->query("//*[@id]");
```

Comment (Progi1984) [reply src/PrestaShopBundle/Install/XmlLoader.php:396]: @PierreRambaud Why using XPath ?

Comment (Progi1984) [reply src/PrestaShopBundle/Install/XmlLoader.php:396]: Fixed.

**Review by PierreRambaud — DISMISSED** [review]

**Review by sowbiba — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,

FO => OK
I found an issue in the BO > Catalog > Brands page
Steps to reproduce the issue:

1. Fresh install "Arabic language" and the United States country
2. Go to BO > Catalog > Brands page > Addresses section
3. The country is displayed in Arabic => ok
4. Change the employee language to "English"
5. Check again the Brands page > Addresses section => Countries are still displayed in Arabic
image

In other pages => ok, for example, Orders page, Addresses page, Countries page
image
image

Thanks to check and feedback.

**Review by atomiix — DISMISSED** [review]

File: classes/Language.php [hunk 13]
```diff
@@ -1434,34 +1435,36 @@ public static function updateMultilangTables(Language $language, array $tablesTo
...
+                INNER JOIN `' . _DB_PREFIX_ . 'country_lang` cl ON c.`id_country` = cl.`id_country`
+                WHERE cl.`id_lang` = "' . (int) $lang->id . '" ';
+        $translatableCountries = Db::getInstance()->executeS($sql, true, false);
+
+        if (empty($translatableCountries)) {
+            return;
+        }
+
+        // Fetch all countries from Intl in specified locale
+        $countries_lang = (new self())->getCountries($lang->locale);
```
Comment (sowbiba) [inline classes/Language.php]:
```suggestion
        $langCountries = (new self())->getCountries($lang->locale);
```

File: src/PrestaShopBundle/Install/LanguageList.php [hunk 14]
```diff
@@ -151,22 +151,34 @@ public function getCountries()
...
+        $countryList = [];
+        $countries_lang = $this->getLanguage($iso)->getCountries();
+        $countries_default = $this->getLanguage(self::DEFAULT_ISO)->getCountries();
+        $xml = @simplexml_load_file(_PS_INSTALL_DATA_PATH_ . 'xml/country.xml');
+        if ($xml) {
+            foreach ($xml->entities->country as $country) {
+                $iso = strtolower((string) $country['iso_code']);
+                $countryList[$iso] = isset($countries_lang[$iso]) ? $countries_lang[$iso] : $countries_default[$iso];
+            }
+        }
```
Comment (sowbiba) [inline src/PrestaShopBundle/Install/LanguageList.php]:
```suggestion
        $langCountries = $this->getLanguage($iso)->getCountries();
        $defaultCountries = $this->getLanguage(self::DEFAULT_ISO)->getCountries();
        $xml = @simplexml_load_file(_PS_INSTALL_DATA_PATH_ . 'xml/country.xml');
        if ($xml) {
            foreach ($xml->entities->country as $country) {
                $iso = strtolower((string) $country['iso_code']);
                $countryList[$iso] = isset($langCountries[$iso]) ? $langCountries[$iso] : $defaultCountries[$iso];
            }
        }
```

**Review by sowbiba — COMMENTED** [review]
Esthetic comments only

**Review by atomiix — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,

It is ok :heavy_check_mark: 
I checked those pages:

1. BO > Orders page / Add new Order page
2. BO > Catalog > Brands Addresses page / Add new Suppliers page
3. BO > Discounts page (country restriction) / Catalog price rule
4. BO > Addresses page
5. BO > Payment method page
6. BO > International > Localisation page (Default country) 
image

6. BO > Geolocalisation page
7. BO > Countries page
8. BO > Tax rules page
9. BO > Stores page
10. FO > Your account > Addresses page
11. FO > Home page > Store information 
12. FO > Checkout > Addresses Step
13. FO > Checkout > Final summary step
14. The Install page 
image

15. The table ps_country_lang => ok

image

Thank you!

**Comment by Progi1984** [comment]
Thanks @khouloudbelguith

**Pull request merged**